### PR TITLE
py-frozendict: New port, version 2.3.10

### PR DESCRIPTION
**Pull request by cquike**
#### Description

This PR adds a new port for the python frozendict module (https://github.com/Marco-Sulla/python-frozendict).

I have tried the API examples shown in the README and they work as expected. I have just tried the python 3.11 port.

###### Type(s)

- [X] submission

###### Tested on
macOS 13.3.1 22E261 arm64

###### Verification 
Have you

- [X] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [X] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [X] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [X] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? 
- [X] checked your Portfile with `port lint --nitpick`?
- [ ] tried existing tests with `sudo port test`?
- [ ] tried a full install with `sudo port -vst install`?
- [X] tested basic functionality of all binary files?
- [X] checked that the Portfile's most important [variants](https://trac.macports.org/wiki/Variants) haven't been broken?

### CODE REVIEW

File: python/py-frozendict/Portfile
Comment (reneeotten) [inline python/py-frozendict/Portfile:5]: please use PyPI i stead of github as the source for Python packages unless really required.

File: python/py-frozendict/Portfile
Comment (reneeotten) [inline python/py-frozendict/Portfile:34]: last tao lines are the default for Python 3.7+ so they can be removed

File: python/py-frozendict/Portfile
Comment (reneeotten) [inline python/py-frozendict/Portfile:14]: this appears to be compiling a C extensions so I doubt this is "noarch"

**Review by reneeotten — CHANGES_REQUESTED** [review]

File: python/py-frozendict/Portfile
Comment (cquike) [reply python/py-frozendict/Portfile]: Sure! I have fixed that now.

File: python/py-frozendict/Portfile
Comment (cquike) [reply python/py-frozendict/Portfile]: Done!

**Review by cquike — COMMENTED** [review]
Thank you for the review! I have tried to fix/answer the issues mentioned.

File: python/py-frozendict/Portfile
Comment (reneeotten) [inline python/py-frozendict/Portfile:27]: distname and master_sites are set by the python PortGroup so these two lines can be removed.

File: python/py-frozendict/Portfile
Comment (reneeotten) [reply python/py-frozendict/Portfile:11]: Looking at the CI it does actually compiles something with clang ; if it is not installed then one should make sure it is also not build.

**Review by reneeotten — CHANGES_REQUESTED** [review]

File: python/py-frozendict/Portfile
Comment (jmroot) [reply python/py-frozendict/Portfile:11]: The CI build is definitely installing the extension.
```
x ./opt/local/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/frozendict/_frozendict.cpython-310-darwin.so
```

File: python/py-frozendict/Portfile
Comment (cquike) [reply python/py-frozendict/Portfile]: I have removed that.

File: python/py-frozendict/Portfile
Comment (cquike) [reply python/py-frozendict/Portfile:11]: Indeed, I see this is being compiled. I am not sure how I managed got an archive without the shared library...

I have fixed that by removing the C code after extracting the tarball. The cleanest way would be to set the env var FROZENDICT_PURE_PY=1 (https://github.com/Marco-Sulla/python-frozendict#building) at build time, but that is not yet supported by the last published tarball. I guess this could be done that way when the next version get's published.

Comment (cquike) [reply python/py-frozendict/Portfile:11]: Now I know why I wasn't getting the C library. I tried with the py311 subport, but apparently there is only C implementation up to 3.10. So the C library was created for the 38, 39 and 310  subports but not for 311 and 312. Mystery solved :)

Comment (jmroot) [reply python/py-frozendict/Portfile:11]: Why would we want to not install the extension where supported?

Comment (cquike) [reply python/py-frozendict/Portfile:11]: Hmm, I thought that the installation from PyPi didn't install the extension, that's why I decided not to install it. But now I realize that PyPi does not install it for python 3.11, but it does for 3.9. So I am not sure what would be the expectation from most users, and whether the extension has been tested enough for all python versions...

Comment (cquike) [reply python/py-frozendict/Portfile:11]: I am more inclined not to ship the extension and wait for user's feedback on the need for the extension. Maybe also when upstream starts supporting 3.12 is a good moment to add the extension, since then all the subports would behave consistently.

Comment (reneeotten) [reply python/py-frozendict/Portfile:11]: Please update to 2.3.9 and enable the C extension. According to the description o. PyPI it should be enabled by default since v2.3.5 and upstream provides .whl files for all Python versions with the compiled extension. Once these changes are made we will merge.

Comment (cquike) [reply python/py-frozendict/Portfile:11]: I have updated it to now enable the C extension. It will be installed for the py38, py39, py310 subports, but not for py311, py312, since frozendict does not provide support for those python versions.
Additionally, I have checkout the v2.3.8 tag from https://github.com/Marco-Sulla/python-frozendict and run the tests with pytest. All the tests pass!
```
 test % python3.12 -m pytest  .
================================================================ test session starts ================================================================
platform darwin -- Python 3.12.0, pytest-7.4.3, pluggy-1.3.0
rootdir: /private/var/tmp/frozendict/python-frozendict
collected 241 items                                                                                                                                 

test_frozendict.py .........................................................................................................................  [ 50%]
test_frozendict_subclass.py ................................................................................................................. [ 97%]
.......                                                                                                                                       [100%]

================================================================ 241 passed in 0.09s ================================================================
 test % python3.11 -m pytest  .
================================================================ test session starts ================================================================
platform darwin -- Python 3.11.6, pytest-7.4.3, pluggy-1.3.0
rootdir: /private/var/tmp/frozendict/python-frozendict
collected 241 items                                                                                                                                 

test_frozendict.py .........................................................................................................................  [ 50%]
test_frozendict_subclass.py ................................................................................................................. [ 97%]
.......                                                                                                                                       [100%]

================================================================ 241 passed in 0.09s ================================================================
 test % python3.10 -m pytest  .
================================================================ test session starts ================================================================
platform darwin -- Python 3.10.13, pytest-7.4.3, pluggy-1.3.0
rootdir: /private/var/tmp/frozendict/python-frozendict
plugins: anyio-3.7.1
collected 482 items                                                                                                                                 

test_frozendict.py .......................................................................................................................... [ 25%]
........................................................................................................................                      [ 50%]
test_frozendict_subclass.py ................................................................................................................. [ 73%]
...............................................................................................................................               [100%]

================================================================ 482 passed in 0.22s ================================================================
 test % python3.9 -m pytest  .
================================================================ test session starts ================================================================
platform darwin -- Python 3.9.18, pytest-7.4.3, pluggy-1.3.0
rootdir: /private/var/tmp/frozendict/python-frozendict
collected 482 items                                                                                                                                 

test_frozendict.py ...............................................................................................sss........................ [ 25%]
........................................................................................................................                      [ 50%]
test_frozendict_subclass.py ..............................................................................................sss................ [ 73%]
...............................................................................................................................               [100%]

========================================================== 476 passed, 6 skipped in 0.22s ===========================================================
 test % python3.8 -m pytest  .
================================================================ test session starts ================================================================
platform darwin -- Python 3.8.18, pytest-7.4.3, pluggy-1.3.0
rootdir: /private/var/tmp/frozendict/python-frozendict
collected 482 items                                                                                                                                 

test_frozendict.py ...............................................................................................sss........................ [ 25%]
........................................................................................................................                      [ 50%]
test_frozendict_subclass.py ..............................................................................................sss................ [ 73%]
...............................................................................................................................               [100%]

========================================================== 476 passed, 6 skipped in 0.23s ===========================================================

```

Comment (cquike) [reply python/py-frozendict/Portfile:11]: Just for confirmation, the contents for the py38, py39, py310 subports contain the C extension but not the ones for py311, py312:

```
user@host frozendict % port contents py38-frozendict | grep .so  
  /opt/local/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/frozendict/_frozendict.cpython-38-darwin.so
user@host frozendict % port contents py39-frozendict | grep .so  
  /opt/local/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/frozendict/_frozendict.cpython-39-darwin.so
user@host frozendict % port contents py310-frozendict | grep .so
  /opt/local/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/frozendict/_frozendict.cpython-310-darwin.so
user@host frozendict % port contents py311-frozendict | grep .so
user@host frozendict % port contents py312-frozendict | grep .so
user@host frozendict % 
```

Comment (reneeotten) [reply python/py-frozendict/Portfile:11]: I don't understand why this is/would be dependent on the Python version. 

If, however, this is the intended behavior then the lines
"supported_archs noarch" and
"platforms {darwin any}"
should be set conditionally only for Python 3.11+

File: python/py-frozendict/Portfile
Comment (reneeotten) [inline python/py-frozendict/Portfile:7]: please update to latest upstream version, which is currently 2.3.10

**Review by reneeotten — CHANGES_REQUESTED** [review]

File: python/py-frozendict/Portfile
Comment (cquike) [reply python/py-frozendict/Portfile:11]: This is because upstream has a different implementation for each python version:
https://github.com/Marco-Sulla/python-frozendict/tree/master/src/frozendict/c_src
So far there is only implementation up to python 3.10. The build system checks whether there is a directory for the python version in question and if there isn't any then it disables automatically the C extension.

**Comment by herbygillot** [comment]
@reneeotten, can you take another look?

**Comment by pmetzger** [comment]
@reneeotten ?

**Pull request merged**
